### PR TITLE
Feature/substreams

### PR DESCRIPTION
**Pull request by maoueh**
Added support for Substreams powered Subgraphs, heavily refactored `src/protocols/index.js` to make it much easier to add new protocols and make it much nicer to read.

### CODE REVIEW

**Comment by azf20** [comment]
❤️

File: src/compiler/index.js [hunk 1]
```diff
@@ -20,36 +23,38 @@ class Compiler {
...
-    for (
-      let dir = path.resolve(this.sourceDir);
-      // Terminate after the root dir or when we have found node_modules
-      dir !== undefined;
-      // Continue with the parent directory, terminate after the root dir
-      dir = path.dirname(dir) === dir ? undefined : path.dirname(dir)
-    ) {
-      if (fs.existsSync(path.join(dir, 'node_modules'))) {
-        this.libsDirs.push(path.join(dir, 'node_modules'))
+    if (options.protocol.name !== 'substreams') {
```
Comment (maoueh) [inline src/compiler/index.js:26]: @dotansimha This was checking `graph-ts` installation, no need for `graph-ts` in Substreams, so removed this check if protocol if `substreams`.

File: src/compiler/index.js [hunk 2]
```diff
@@ -243,7 +252,11 @@ class Compiler {
     )
   }
 
-  _compileDataSourceMapping(dataSource, mappingPath, compiledFiles, spinner) {
+  _compileDataSourceMapping(protocol, dataSource, mappingPath, compiledFiles, spinner) {
+    if (protocol.name == 'substreams') {
```
Comment (maoueh) [inline src/compiler/index.js:256]: @dotansimha Skipped compilation steps if not required, which is the case for `Substreams`

File: src/compiler/index.js [hunk 3]
```diff
@@ -437,6 +450,30 @@ class Compiler {
                 )
             }
 
+            if (protocol.name == 'substreams') {
```
Comment (maoueh) [inline src/compiler/index.js:453]: This resolves `.spkg` and copy it to `build` directory, when on `Substreams`

Comment (maoueh) [reply src/compiler/index.js:453]: @dotansimha

File: src/compiler/index.js [hunk 4]
```diff
@@ -537,15 +574,28 @@ class Compiler {
...
-          updates.push({
-            keyPath: ['dataSources', i, 'mapping', 'file'],
-            value: await this._uploadFileToIPFS(
-              dataSource.getIn(['mapping', 'file']),
-              uploadedFiles,
-              spinner,
-            ),
-          })
+        if (this.protocol.name !== 'substreams') {
+          for (let [i, dataSource] of subgraph.get('dataSources').entries()) {
```
Comment (maoueh) [inline src/compiler/index.js:578]: @dotansimha If **not** Substreams, perform previous logic which is to upload and resolve WASM file to IPFS.

File: src/compiler/index.js [hunk 5]
```diff
@@ -537,15 +574,28 @@ class Compiler {
...
+              keyPath: ['dataSources', i, 'mapping', 'file'],
+              value: await this._uploadFileToIPFS(
+                dataSource.getIn(['mapping', 'file']),
+                uploadedFiles,
+                spinner,
+              ),
+            })
+          }
+        } else {
+          for (let [i, dataSource] of subgraph.get('dataSources').entries()) {
```
Comment (maoueh) [inline src/compiler/index.js:589]: @dotansimha If Substreams, perform new logic which is to upload and resolve `.spkg` file to IPFS.

File: src/protocols/index.js [hunk 6]
```diff
@@ -6,24 +6,51 @@ const EthereumABI = require('./ethereum/abi')
 const EthereumSubgraph = require('./ethereum/subgraph')
```
Comment (maoueh) [inline src/protocols/index.js:6]: Heavily refactored file, https://github.com/graphprotocol/graph-cli/pull/991/commits/303f0c8941787e1201204ba3f3a05ac9c58f4bec has been made to isolate this change from the rest of the PR. 

Make it much simpler to add new networks. And I spotted a mistake along the way:

```
  getManifestScaffold() {
    switch (this.name) {
      case 'arweave':
        return ArweaveMappingScaffold
      case 'ethereum':
        return EthereumManifestScaffold
      case 'near':
        return NearManifestScaffold
      case 'cosmos':
        return CosmosManifestScaffold
    }
  }
        ```
 
 The `ArweaveMappingScaffold` is wrong, should have been `ArweaveManifestScaffold`

File: src/protocols/substreams/manifest.graphql [hunk 7]
```diff
@@ -0,0 +1,45 @@
+# Each referenced type's in any of the types below must be listed
```
Comment (maoueh) [inline src/protocols/substreams/manifest.graphql:1]: Added to validate Substreams's Subgraph manifest format

File: src/subgraph.js [hunk 8]
```diff
@@ -134,11 +137,19 @@ Please update it to tell users more about your subgraph.`,
...
-
         const handlerTypes = protocolSubgraph.handlerTypes()
 
+        subgraphDebug(
+          'Validating dataSource "%s" handlers with %d handlers types defined for protocol',
+          dataSource.get('name'),
+          handlerTypes.size,
+        )
+
+        if (handlerTypes.size == 0) {
```
Comment (maoueh) [inline src/subgraph.js:149]: @dotansimha  Substreams has no handlers, so needed to add that check to return earlier.

File: src/compiler/index.js (hunk 3 above)
Comment (dotansimha) [reply src/compiler/index.js:453]: thanks!

File: src/protocols/index.js (hunk 6 above)
Comment (dotansimha) [reply src/protocols/index.js:6]: I think ideally, we'll need to eventually to make this code more generic and have the different chain hooked through an external piece of code.

**Comment by leoyvens** [comment]
This lgtm

**Comment by maoueh** [comment]
By the way, I'm able to reproduce the CI complaining about `Cannot redefine property: wrap` by running:

```
yarn test --runInBand --detectOpenHandles
```

Which points also to the leaked module:

```
Jest has detected the following 1 open handle potentially keeping Jest from exiting:

  ●  PROCESSWRAP

      at Process.emitInitNative (node:internal/async_hooks:205:43)
      at new ChildProcess (node:internal/child_process:263:18)
      at spawn (node:child_process:695:17)
      at start (node_modules/sync-rpc/lib/index.js:33:13)
      at sendMessage (node_modules/sync-rpc/lib/index.js:133:17)
      at createClient (node_modules/sync-rpc/lib/index.js:173:27)
      at Object.<anonymous> (node_modules/sync-request/lib/index.js:16:14)
```

**Comment by maoueh** [comment]
Rebased, please merge, I don't have the rights to do so

**Comment by maoueh** [comment]
Fix also a small problem in `graph codegen` where `--skipMigrations` was not in the list of fix parameters that can have no argument: https://github.com/graphprotocol/graph-cli/pull/991/files#diff-0e7b48d459a4312959b62780d78da0af2b8948320e1aec542f8aeab817bd0ef0R59

**Comment by maoueh** [comment]
Sorry seems my comment about CI failing was not attached here. The CI is failing but it does not seem because of this PR, seems something that is there for a long time.

Locally all the tests pass, the error is:

```
TypeError: Cannot redefine property: wrap
```

Which I read on the internet that it's caused by `gluegun` and having multiple instances of the the assemblyscript compiler or something like that, see for example https://github.com/AssemblyScript/assemblyscript/issues/1948.

I would merge the PR nonetheless and hopefully while refactoring you are able to determine what is wrong with the CI exactly.

The way I found to reproduce locally is using `yarn test --runInBand --detectOpenHandles`, it seems it triggers the error, again, unclear to me where it is exactly.

**Comment by evaporei** [comment]
I just published this in `0.37.0` :wink: :tada: 
Thanks for the PR!! :rocket:

**Pull request merged**